tunnel/intra/doh/doh_test.go

```diff
@@ -559,3 +684,134 @@ func TestAcceptOversize(t *testing.T) {
 		t.Error("Expected to read 0 bytes")
 	}
 }
+
+func TestComputePaddingSize(t *testing.T) {
+	if computePaddingSize(100-kOptPaddingHeaderLen, 100) != 0 {
+		t.Errorf("Expected no padding")
+	}
+	if computePaddingSize(200-kOptPaddingHeaderLen, 100) != 0 {
+		t.Errorf("Expected no padding")
+	}
+	if computePaddingSize(190-kOptPaddingHeaderLen, 100) != 10 {
+		t.Errorf("Expected to pad up to next block")
+	}
+}
+
+func TestAddEdnsPaddingIdempotent(t *testing.T) {
+	padded, err := AddEdnsPadding(simpleQueryBytes)
+	if err != nil {
+		panic(err)
+	}
+	paddedAgain, err := AddEdnsPadding(padded)
+	if err != nil {
+		panic(err)
+	}
+	if !bytes.Equal(padded, paddedAgain) {
+		t.Errorf("Padding should be idempotent\n%v\n%v", padded, paddedAgain)
+	}
+}
+
+// Sanity check that packing |compressedQueryBytes| constructs the same query
+// byte-for-byte.
+func TestDnsMessageCompressedQuerySanityCheck(t *testing.T) {
+	m := mustUnpack(compressedQueryBytes)
+	packedBytes := mustPack(m)
+	if len(packedBytes) != len(compressedQueryBytes) {
+		t.Errorf("Packed query has different size than original:\n  %v\n  %v", packedBytes, compressedQueryBytes)
+	}
+}
+
+// Sanity check that packing |uncompressedQueryBytes| constructs a smaller
+// query byte-for-byte, since label compression is enabled by default.
+func TestDnsMessageUncompressedQuerySanityCheck(t *testing.T) {
+	m := mustUnpack(uncompressedQueryBytes)
+	packedBytes := mustPack(m)
+	if len(packedBytes) >= len(uncompressedQueryBytes) {
+		t.Errorf("Compressed query is not smaller than uncompressed query")
+	}
+}
+
+// Check that we correctly pad an uncompressed query to the nearest block.
+func TestAddEdnsPaddingUncompressedQuery(t *testing.T) {
+	if len(uncompressedQueryBytes)%PaddingBlockSize == 0 {
+		t.Errorf("uncompressedQueryBytes does not require padding, so this test is invalid")
+	}
+	padded, err := AddEdnsPadding(uncompressedQueryBytes)
+	if err != nil {
+		panic(err)
+	}
+	if len(padded)%PaddingBlockSize != 0 {
+		t.Errorf("AddEdnsPadding failed to correctly pad uncompressed query")
+	}
+}
+
+// Check that we correctly pad a compressed query to the nearest block.
+func TestAddEdnsPaddingCompressedQuery(t *testing.T) {
+	if len(compressedQueryBytes)%PaddingBlockSize == 0 {
+		t.Errorf("compressedQueryBytes does not require padding, so this test is invalid")
+	}
+	padded, err := AddEdnsPadding(compressedQueryBytes)
+	if err != nil {
+		panic(err)
+	}
+	if len(padded)%PaddingBlockSize != 0 {
+		t.Errorf("AddEdnsPadding failed to correctly pad compressed query")
+	}
+}
+
+// Try to pad a query that already contains an OPT record, but no padding option.
+func TestAddEdnsPaddingCompressedOptQuery(t *testing.T) {
+	optQuery := simpleQuery
+	optQuery.Additionals = append(optQuery.Additionals,
```

Comment:
I think this is permitted to mutate `simpleQuery.Additionals`.  It might not actually be mutated due to the way the slice is allocated, but I would prefer an explicit `copy` just to be safe.

Reply:
OK, now I'm explicitly copying `simpleQuery.Additionals` before appending to it.